authorization/invitation/model/invitation_model_service.go

```diff
@@ -0,0 +1,221 @@
+		// If the UserID has been provided, confirm it is valid and that the identity is a user
+		if invitation.IdentityID != nil {
+			identity, err := s.repo.Identities().Load(ctx, *invitation.IdentityID)
+			if err != nil {
+				return errors.NewInternalErrorFromString(ctx, fmt.Sprintf("invalid identity ID specified: %s\n", invitation.IdentityID))
+			}
+
+			if !identity.IsUser() {
+				return errors.NewBadParameterErrorFromString("Identity ID", invitation.IdentityID, "identity is not a user")
+			}
+		} else if invitation.UserName != nil {
+			// If the username has been provided, confirm the user is valid and that the identity is a user, and set the UserID
+			identities, err := s.repo.Identities().Query(account.IdentityFilterByUsername(*invitation.UserName))
+			if err != nil {
+				return errors.NewInternalError(ctx, err)
+			}
+			// If there is anything other than 1 result found, we have a problem
+			if len(identities) == 0 {
+				// If no users are found, return an error
+				return errors.NewBadParameterErrorFromString("Username", invitation.UserName, "username not found")
+			} else if len(identities) > 1 {
+				// If more than one user is found, return an error
+				return errors.NewBadParameterErrorFromString("Username", invitation.UserName, "more than one user with same username found")
+			}
+
+			// Set the IdentityID to that of the identity found
+			invitations[i].IdentityID = &identities[0].ID
+		} else if invitation.UserEmail != nil {
+			// If the user's e-mail address has been provided, confirm the user is valid and that the identity is a user, and set the UserID
+			users, err := s.repo.Users().Query(account.UserFilterByEmail(*invitation.UserEmail))
+			if err != nil {
+				return errors.NewInternalError(ctx, err)
+			}
+			// We expect exactly 1 user to be found, if not we return an error
+			if len(users) == 0 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "user with e-mail address not found")
+			} else if len(users) > 1 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "more than one user with e-mail address found")
+			}
+
+			userID := &users[0].ID
+
+			// Now that we have the user ID, we can lookup the identity ID
+			identities, err := s.repo.Identities().Query(account.IdentityFilterByUserID(*userID))
+			// If there is anything other than 1 result found, return an error
+			if len(identities) == 0 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "no identity found for user with e-mail address")
+			} else if len(identities) > 1 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "more than one identity found for user with e-mail address")
+			}
+
+			// Set the IdentityID to that of the identity found
+			invitations[i].IdentityID = &identities[0].ID
+		}
+
+		// Confirm that any specified roles are valid for this resource type
+		for _, roleName := range invitation.Roles {
```

Comment:
I've removed this section.  The point of it was to validate the role names before we started creating records, but I guess that since this entire thing is in a transaction nothing will get created if the transaction fails.